src/js/components/JSONEditor.js

```diff
@@ -0,0 +1,504 @@
+      mutations.forEach((mutation) => {
+        if (!mutation.addedNodes) {
+          return;
+        }
+        // Wait until we the tooltip DOM element is injected and position it
+        // accordingly
+        mutation.addedNodes.forEach((node) => {
+          if (node.classList.contains('ace_tooltip')) {
+            this.handleWindowResize();
+            observer.disconnect();
+          }
+        });
+      });
+    });
+    observer.observe(this.aceEditor.container, {childList: true});
+    //
+    // -- END HACK -------------------------------------------------------------
+
+  }
+
+  /**
+   * `onBlur` handler for the AceEditor
+   *
+   * @param {DOMEvent} event - The event object
+   */
+  handleBlur(event) {
+    this.props.onBlur(event, this.jsonValue);
+    this.isFocused = false;
+  }
+
+  /**
+   * `onChange` handler for the AceEditor
+   *
+   * @param {string} jsonText - The new JSON string
+   */
+  handleChange(jsonText) {
+    let lastValue = this.jsonValue;
+    let lastError = this.jsonError;
+
+    // Calculate what the next state is going to be
+    this.updateLocalJsonState(jsonText);
+
+    // Handle errors
+    if (lastError !== this.jsonError) {
+      this.props.onErrorStateChange(this.jsonError);
+    }
+
+    // Update the `isTyping` flag
+    this.isTyping = true;
+    this.scheduleIsTypingReset(ISTYPING_TIMEOUT);
+
+    // If we have errors don't continue with updating the local structures
+    if (this.jsonError) {
+      return;
+    }
+
+    // Calculate differences in the JSON and trigger `onPropertyChange`
+    // event for every property that changed in the JSON
+    let diff = JSONEditorUtil.deepObjectDiff(lastValue,
+      this.jsonValue);
+    diff.forEach((diff) => {
+      this.props.onPropertyChange(diff.path, diff.value,
+        this.jsonValue);
+    });
+
+    // Trigger change with the latest json object
+    this.props.onChange(this.jsonValue);
+  }
+
+  /**
+   * `onFocus` handler for the AceEditor
+   *
+   * @param {DOMEvent} event - The event object
+   */
+  handleFocus(event) {
+    this.props.onFocus(event, this.jsonValue);
+    this.isFocused = true;
+  }
+
+  /**
+   * Apply pending state updates to the ACE editor instance.
+   *
+   * This function operates purely on the `aceEditor` instance, created by the
+   * `AceEditor` composed and exposed through the `onLoad` event. Since the
+   * `AceEditor` component does a pretty bad job updating only what's needed,
+   * we had to use this approach and avoid re-rendering the component as much
+   * as possible.
+   */
+  updateEditorState() {
+    if (!this.aceEditor) {
+      return;
+    }
+
+    // Prepare markers
+    let markers = [];
+
+    // Extract syntax errors, or other errors that refer to line
+    if (this.jsonError) {
+
+      // Strip out the 'at line xxx' message, and keep track of that line
+      let errorLine = 0;
+      let errorMsg = this.jsonError.replace(/at line ([\d:]+)/g,
```

Comment:
After we have decided to use big arrow functions, we can push this up into the line above